worker/machineactions/worker.go

```diff
@@ -103,30 +109,59 @@ func (h *handler) Handle(_ <-chan struct{}, actionsSlice []string) error {
 		if err != nil {
 			return errors.Annotatef(err, "could not retrieve action %s", actionId)
 		}
-
-		err = h.config.Facade.ActionBegin(actionTag)
-		if err != nil {
-			return errors.Annotatef(err, "could not begin action %s", action.Name())
-		}
-
-		// We try to handle the action. The result returned from handling the action is
-		// sent through using ActionFinish. We only stop the loop if ActionFinish fails.
-		var finishErr error
-		results, err := h.config.HandleAction(action.Name(), action.Params(), action.Parallel(), action.ExecutionGroup())
-		if err != nil {
-			finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionFailed, nil, err.Error())
-		} else {
-			finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionCompleted, results, "")
-		}
-		if finishErr != nil {
-			return errors.Trace(finishErr)
-		}
+		// TODO(wallyworld) - use a thread pool to put a cap on the max number of parallel actions.
+		h.wait.Add(1)
+		go func(action machineactions.Action) {
+			var results map[string]interface{}
+			var actionErr error
+			defer func() {
+				// The result returned from handling the action is sent through using ActionFinish.
+				var finishErr error
+				if actionErr != nil {
+					finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionFailed, nil, actionErr.Error())
+				} else {
+					finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionCompleted, results, "")
+				}
+				if finishErr != nil {
+					logger.Errorf("could not finish action %s: %v", action.Name(), finishErr)
+				}
+			}()
+
+			defer h.wait.Done()
+			if !action.Parallel() || action.ExecutionGroup() != "" {
+				group := "exec-command"
+				worker := "machine exec command runner"
+				if g := action.ExecutionGroup(); g != "" {
+					group = fmt.Sprintf("%s-%s", group, g)
+					worker = fmt.Sprintf("%s (exec group=%s)", worker, g)
+				}
+				spec := machinelock.Spec{
+					Cancel:  abort,
+					Worker:  worker,
+					Comment: fmt.Sprintf("action %s", action.ID()),
+					Group:   group,
+				}
+				releaser, err := h.config.MachineLock.Acquire(spec)
```

Comment:
Is there a timeout waiting for this lock, or is it indefinite? I guess indefinite makes sense in the context of running arbitrary actions?

Reply:
Yeah, we don't set a timeout which means the underlying mutex waits indefinitely.